lib/galaxy/datatypes/gis.py

```
@@ -0,0 +1,123 @@
+            else:
+                rval.append('<li><a href="%s" type="application/binary">%s</a>%s</li>' % (fn, fn, opt_text))
+        rval.append('</ul></div></html>')
+        return "\n".join(rval)
+
+    def regenerate_primary_file(self, dataset):
+        """
+        cannot do this until we are setting metadata
+        """
+        shp = dataset.extra_files_path
+        flist = os.listdir(shp)
+        rval = ['<html><head><title>Files for Composite Dataset %s</title></head><body><p/>Composite %s contains:<p/><ul>' % (dataset.name, dataset.name)]
+        for i, fname in enumerate(flist):
+            sfname = os.path.split(fname)[-1]
+            f, e = os.path.splitext(fname)
+            rval.append('<li><a href="%s">%s</a></li>' % (sfname, sfname))
+        rval.append('</ul></body></html>')
+        with open(dataset.file_name, 'w') as f:
+            f.write("\n".join(rval))
+            f.write('\n')
+
+    def set_meta(self, dataset, **kwd):
+        """
+
+        """
+        GIS.set_meta(self, dataset, **kwd)
+        if not kwd.get('overwrite'):
+            if verbose:
+                gal_Log.debug('@@@ shapefile set_meta called with overwrite = False')
+            return True
+        try:
+            shp = dataset.extra_files_path
+        except Exception:
+            if verbose:
+                gal_Log.debug('@@@shapefile set_meta failed %s - dataset %s has no shp ?' % (sys.exc_info()[0], dataset.name))
+            return False
+        try:
+            flist = os.listdir(shp)
+        except Exception:
+            if verbose:
+                gal_Log.debug('@@@shapefile set_meta failed %s - dataset %s has no shp ?' % (sys.exc_info()[0], dataset.name))
+            return False
+        if len(flist) == 0:
+            if verbose:
+                gal_Log.debug('@@@shapefile set_meta failed - %s shp %s is empty?' % (dataset.name, shp))
+            return False
```

If it failed you should raise an Exception here.